src/main/kotlin/cacophony/controlflow/FunctionHandler.kt

```diff
@@ -271,54 +242,97 @@ class PrologueEpilogueHandler(
+
+    private val prologueEpilogueHandler =
+        PrologueEpilogueHandler(
+            this,
+            callConvention,
+            stackSpace,
+            resultRegister,
+        )
+
+    override fun generatePrologue(): List<CFGNode> = prologueEpilogueHandler.generatePrologue()
+
+    override fun generateEpilogue(): List<CFGNode> = prologueEpilogueHandler.generateEpilogue()
+
+    // Creates staticLink auxVariable in analyzedFunction, therefore shouldn't be called multiple times.
+    // Static link is created even if parent doesn't exist.
+    private fun introduceStaticLinksParams() {
+        registerVariableAllocation(
+            staticLink,
+            VariableAllocation.OnStack(0),
+        )
+        analyzedFunction.auxVariables.add(staticLink)
+    }
+}
+
+class PrologueEpilogueHandler(
+    private val handler: FunctionHandler,
+    private val callConvention: CallConvention,
+    private val stackSpace: CFGNode.ConstantLazy,
+    private val resultAccess: Register.VirtualRegister,
+) {
+    private val spaceForPreservedRegisters: List<Register.VirtualRegister> =
+        callConvention.preservedRegisters().map {
+            Register.VirtualRegister()
         }
 
     fun generatePrologue(): List<CFGNode> {
         val nodes = mutableListOf<CFGNode>()
         with(handler) {
-            // Move first, then change RSP, so that the offsets from the callConvention hold
-            // However, keep the existence of red zone in mind
+            nodes.add(pushRegister(rbp))
+            nodes.add(registerUse(rbp) assign (registerUse(rsp) sub CFGNode.ConstantReal(REGISTER_SIZE)))
+            nodes.add(registerUse(rsp) subeq stackSpace)
+
+            // Preserved registers
+            for ((source, destination) in callConvention.preservedRegisters() zip spaceForPreservedRegisters) {
+                nodes.add(registerUse(destination) assign registerUse(Register.FixedRegister(source)))
+            }
 
             // Defined function arguments
-            for ((ind, arg) in function.arguments.withIndex()) {
+            for ((ind, arg) in getFunctionDeclaration().arguments.withIndex()) {
                 nodes.add(
-                    lvalueFromAllocation(getVariableAllocation(getVariableFromDefinition(arg))) assign
-                        callConvention(ind),
+                    wrapAllocation(getVariableAllocation(getVariableFromDefinition(arg))) assign
+                        wrapAllocation(callConvention.argumentAllocation(ind)),
                 )
             }
+
             // Static link (implicit arg)
             nodes.add(
-                lvalueFromAllocation(getVariableAllocation(getStaticLink())) assign
-                    callConvention(function.arguments.size),
+                wrapAllocation(getVariableAllocation(getStaticLink())) assign
+                    wrapAllocation(callConvention.argumentAllocation(getFunctionDeclaration().arguments.size)),
```

Comment:
Currently the static link is the first parameter, not last (see line 144). I think it would be nice to have it specified in the calling convention object, but it is not necessary.

Reply:
Oh, really? I thought we have it as the last argument, so that calling our functions from C/C++ and vice versa is less cumbersome

Reply:
Unfortunately yes. You can change it now, but I think we will do it regardless once we implement FFI.

Reply:
I will do it in another PR